src/pymor/analyticalproblems/functions.py

```diff
@@ -380,3 +381,130 @@ def evaluate(self, x, mu=None):
+        assert evaluation_points is None or isinstance(evaluation_points, np.ndarray) and \
+            evaluation_points.ndim == 2 and evaluation_points.shape[1] == function.dim_domain
+        assert basis_evaluations is None or isinstance(basis_evaluations, np.ndarray) and \
+            basis_evaluations.shape == (len(interpolation_points), len(evaluation_points)) + function.shape_range
+
+        self.__auto_init(locals())
+        functions = [EmpiricalInterpolatedFunctionBasisFunction(self, i) for i in range(len(interpolation_points))]
+        coefficients = [EmpiricalInterpolatedFunctionFunctional(self, i) for i in range(len(interpolation_points))]
+        super().__init__(functions, coefficients)
+        self._last_evaluation_points, self._last_basis_evaluations = evaluation_points, basis_evaluations
+        if self._last_evaluation_points is not None:
+            self._last_evaluation_points.flags.writeable = False
+            self._last_basis_evaluations.flags.writeable = False
+        self._last_mu = None
+
+    def _update_coefficients(self, mu):
+        if mu is not self._last_mu:
+            assert self.parameters.assert_compatible(mu)
+            self._last_mu = mu
+            fx = self.function.evaluate(self.interpolation_points, mu=mu)
+            if self.triangular:
+                self._last_interpolation_coefficients = solve_triangular(
+                    self.interpolation_matrix, fx, lower=True, unit_diagonal=True
+                )
+            else:
+                self._last_interpolation_coefficients = solve(self.interpolation_matrix, fx)
+
+    def _update_evaluation_points(self, x):
+        if self._last_evaluation_points is not None and np.all(x == self._last_evaluation_points):
+            return
+        if self._last_evaluation_points is not None:
+            self.logger.info('Evaluating function snapshots for new evaluation points.')
+        x.flags.writeable = False
+        self._last_evaluation_points = x
+        snapshot_evaluations = np.array([self.function.evaluate(x, mu=mu) for mu in self.snapshot_mus])
+        self._last_basis_evaluations = self.snapshot_coefficients @ snapshot_evaluations.T
+
+    def evaluate(self, x, mu=None):
+        self._update_evaluation_points(x)
+        basis_evaluations = self._last_basis_evaluations
+        interpolation_coefficients = self._evaluate_coefficients(mu)
+        return basis_evaluations.T @ interpolation_coefficients
+
+
+class EmpiricalInterpolatedFunctionFunctional(ParameterFunctional):
+
+    def __init__(self, interpolated_function, index):
+        self.__auto_init(locals())
+        # we explicitly have to set the parameters since interpolation_points isn't initialized yet
+        self.parameters = interpolated_function.function.parameters
+
+    def evaluate(self, mu=None):
+        self.interpolated_function._update_coefficients(mu)
+        return self.interpolated_function._last_interpolation_coefficients[self.index]
```

Comment:
When doing empirical interpolation of e.g. a non-parametric function, `_last_interpolation_coefficients` will be undefined since `self.interpolated_function._update_coefficients(mu)` will do nothing. However, I am not sure if this case is even relevant.

Reply:
Fixed by setting `_last_mu` to `'NONE'`.